File: apps/api/services/director/analysis/plan_chunks.py
"""Long-form analysis chunk planning (Phase 12)."""
from __future__ import annotations

from dataclasses import dataclass

CHUNK_THRESHOLD_SECONDS = 15 * 60
DEFAULT_CHUNK_TARGET_MINUTES = 9
DEFAULT_OVERLAP_SECONDS = 25


@dataclass(frozen=True)
class ChunkPlan:
    chunk_index: int
    core_start: float
    core_end: float
    window_start: float
    window_end: float

    def to_dict(self) -> dict:
        return {
            "chunkIndex": self.chunk_index,
            "coreStart": self.core_start,
            "coreEnd": self.core_end,
            "windowStart": self.window_start,
            "windowEnd": self.window_end,
        }


def plan_chunks(
    duration_seconds: float,
    *,
    chunk_target_minutes: int = DEFAULT_CHUNK_TARGET_MINUTES,
    overlap_seconds: int = DEFAULT_OVERLAP_SECONDS,
    chunk_threshold_seconds: int = CHUNK_THRESHOLD_SECONDS,
) -> list[ChunkPlan]:
    """Plan overlapping analysis windows for long-form content."""
    duration = max(0.0, float(duration_seconds))
    if duration == 0:
        return [
            ChunkPlan(0, 0.0, 0.0, 0.0, 0.0),
        ]

    if duration <= chunk_threshold_seconds:
        return [
            ChunkPlan(0, 0.0, duration, 0.0, duration),
        ]

    core_length = chunk_target_minutes * 60
    overlap = float(overlap_seconds)
    chunks: list[ChunkPlan] = []
    core_start = 0.0
    chunk_index = 0

    while core_start < duration:
        core_end = min(core_start + core_length, duration)
        chunks.append(
            ChunkPlan(
                chunk_index=chunk_index,
                core_start=core_start,
                core_end=core_end,
                window_start=max(0.0, core_start - overlap),
                window_end=min(duration, core_end + overlap),
            )
        )
        core_start = core_end
        chunk_index += 1

    return chunks


def overlap_zone(left: ChunkPlan, right: ChunkPlan) -> tuple[float, float] | None:
    """Return overlap window between adjacent chunk plans."""
    start = max(left.window_start, right.window_start)
    end = min(left.window_end, right.window_end)
    if end <= start:
        return None
    return start, end

File: remotion-service/src/lib/analysis/planChunks.ts
/**
 * Long-form analysis chunk planning (Phase 12).
 *
 * Below CHUNK_THRESHOLD_SECONDS: single no-op chunk — short-form behavior unchanged.
 * Above threshold: ~8–10 minute core windows with ~20–30 second overlap on each side.
 */

export const CHUNK_THRESHOLD_SECONDS = 15 * 60;
export const DEFAULT_CHUNK_TARGET_MINUTES = 9;
export const DEFAULT_OVERLAP_SECONDS = 25;

export interface ChunkPlan {
  chunkIndex: number;
  /** Non-overlapping owned region start (seconds). */
  coreStart: number;
  /** Non-overlapping owned region end (seconds). */
  coreEnd: number;
  /** coreStart minus overlap, clamped to 0. */
  windowStart: number;
  /** coreEnd plus overlap, clamped to file duration. */
  windowEnd: number;
}

export interface PlanChunksOptions {
  chunkTargetMinutes?: number;
  overlapSeconds?: number;
  chunkThresholdSeconds?: number;
}

export function planChunks(
  durationSeconds: number,
  options?: PlanChunksOptions,
): ChunkPlan[] {
  const duration = Math.max(0, durationSeconds);
  if (duration === 0) {
    return [
      {
        chunkIndex: 0,
        coreStart: 0,
        coreEnd: 0,
        windowStart: 0,
        windowEnd: 0,
      },
    ];
  }

  const threshold =
    options?.chunkThresholdSeconds ?? CHUNK_THRESHOLD_SECONDS;
  if (duration <= threshold) {
    return [
      {
        chunkIndex: 0,
        coreStart: 0,
        coreEnd: duration,
        windowStart: 0,
        windowEnd: duration,
      },
    ];
  }

  const coreLength =
    (options?.chunkTargetMinutes ?? DEFAULT_CHUNK_TARGET_MINUTES) * 60;
  const overlap = options?.overlapSeconds ?? DEFAULT_OVERLAP_SECONDS;
  const chunks: ChunkPlan[] = [];
  let coreStart = 0;
  let chunkIndex = 0;

  while (coreStart < duration) {
    const coreEnd = Math.min(coreStart + coreLength, duration);
    chunks.push({
      chunkIndex,
      coreStart,
      coreEnd,
      windowStart: Math.max(0, coreStart - overlap),
      windowEnd: Math.min(duration, coreEnd + overlap),
    });
    coreStart = coreEnd;
    chunkIndex += 1;
  }

  return chunks;
}

/** Overlap zone between adjacent chunk windows (for reconciliation). */
export function overlapZone(
  left: ChunkPlan,
  right: ChunkPlan,
): { start: number; end: number } | null {
  const start = Math.max(left.windowStart, right.windowStart);
  const end = Math.min(left.windowEnd, right.windowEnd);
  if (end <= start) {
    return null;
  }
  return { start, end };
}
